cmake/Boost.cmake

```diff
@@ -0,0 +1,93 @@
+macro(p4c_obtain_boost)
+  option(
+    P4C_USE_PREINSTALLED_BOOST
+    "Look for a preinstalled version of Boost in the system instead of installing the library using FetchContent."
+    OFF
+  )
+
+  # If P4C_USE_PREINSTALLED_BOOST is ON just try to find a preinstalled version of Boost.
+  if(P4C_USE_PREINSTALLED_BOOST)
+    if(NOT BUILD_SHARED_LIBS)
+      # Link Boost statically.
+      # See https://cmake.org/cmake/help/latest/module/FindBoost.html for details.
+      set(Boost_USE_STATIC_LIBS ON)
+      set(Boost_USE_STATIC_RUNTIME OFF)
+    endif()
+
+    # The boost graph headers are optional and only required by the graphs back end.
```

Comment:
It might be a tradeoff between moving a bunch of boost logic to the top-level or keeping it all contained within a single CMake file. I pushed some changes which explore the other option.

Not sure which way is preferable.